### PR TITLE
Allows users to customize artifact names

### PR DESCRIPTION
**Pull request by kenxu95**
## Describe your changes and why you are making these changes

Allows for both:
```
artf = fn()
artf.set_name('featurized data')
```

And:
```
@op(outputs=["featurized data"])
def foo():
    ...
```

## Checklist before requesting a review
- [ ] I have created a descriptive PR title. The PR title should complete the sentence "This PR...".
- [ ] I have performed a self-review of my code.
- [ ] I have included a small demo of the changes. For the UI, this would be a screenshot or a Loom video.
- [ ] If this is a new feature, I have added unit tests and integration tests.
- [ ] I have run the integration tests locally and they are passing.
- [ ] I have run the linter script locally (See `python3 scripts/run_linters.py -h` for usage).
- [ ] All features on the UI continue to work correctly.
- [ ] Added one of the following CI labels:
    - `run_integration_test`: Runs integration tests
    - `skip_integration_test`: Skips integration tests (Should be used when changes are ONLY documentation/UI)

### CODE REVIEW

**Comment by vsreekanti** [comment]
API looks good! Just to double check, how'd we end up dealing with multiple outputs in the decorator?

**Comment by kenxu95** [comment]
If there are multiple outputs, you either have to name all of them:
```
@op(outputs=["output1", "output2"])
def foo():
    ...
```

or none

```
@op(num_outputs=2)
def foo():
    ...
```

or you can do both, but it's unnecessary:

```
@op(num_outputs=2, outputs=["output1", "output2"])
def foo():
    ...
```

**Comment by vsreekanti** [comment]
Okay, that's great!

File: sdk/aqueduct/decorator.py [hunk 1]
```diff
@@ -373,15 +407,25 @@ def op(
...
+    if num_outputs is None and outputs is None:
+        num_outputs = 1
+    elif num_outputs is not None and outputs is not None:
+        if len(outputs) != num_outputs:
+            raise InvalidUserArgumentException(
+                "`len(outputs) must be equivalent to `num_outputs`. Getting %d and %d"
+                % (len(outputs), num_outputs)
+            )
+    elif num_outputs is None and outputs is not None:
+        num_outputs = len(outputs)
```
Comment (eunice-chan) [inline sdk/aqueduct/decorator.py:420]: Will there be any issue if len(outputs) != actual num_outputs?

Comment (kenxu95) [reply sdk/aqueduct/decorator.py:420]: yep, it'll fail at execution time.

File: sdk/aqueduct/decorator.py [hunk 2]
```diff
@@ -373,15 +407,25 @@ def op(
...
+
+    # If not set, default number of outputs is one.
```
Comment (likawind) [inline sdk/aqueduct/decorator.py:422]: I'm a bit confused as after the above 3 if-elif blocks the only remaining condition is `num_output is not None` and outputs is None` ? It seems that we should do some assertion here instead?

**Review by likawind — APPROVED** [review]
The implementation and testings looks good! Just one minor comments.

I think the change does reflect the discussions in `product` channel, but will let Vikram confirm on API pattern.

File: sdk/aqueduct/decorator.py (hunk 2 above)
Comment (kenxu95) [reply sdk/aqueduct/decorator.py:422]: Yeah the casework here is a little tricky - if a `num_output` is provided, then the `outputs` is an optional argument, hence why that case isn't specified here. Open to suggestions, but the relationship isn't symmetric

**Pull request merged**